leasing/report/invoice/invoicing_review.py

```diff
@@ -507,8 +507,32 @@ def data_as_excel(self, data_sections):
             FormatType.AREA: workbook.add_format({"num_format": r"#,##0.00 \m\²"}),
         }
 
+        # worksheet max length name is 31 so need to truncate
+        worksheet_names_with_duplicates = map(
+            lambda key: (key[:29] + "..") if len(key) > 31 else key,
+            data_sections.keys(),
+        )
+
+        # need to make sure there are no duplacted worksheet names so add numbering to end if duplicates
+        duplicate_name_count_dict = {}
+        worksheet_names = []
+        for worksheet_name in worksheet_names_with_duplicates:
+            if worksheet_name in duplicate_name_count_dict:
+                duplicate_name_count_dict[worksheet_name] += 1
+                str_num_to_append = str(duplicate_name_count_dict[worksheet_name])
+                worksheet_names.append(
+                    worksheet_name[: -(1 + len(str_num_to_append))]
+                    + "_"
+                    + str(duplicate_name_count_dict[worksheet_name])
+                )
+            else:
+                duplicate_name_count_dict[worksheet_name] = 1
+                worksheet_names.append(worksheet_name)
+
         for index, (key, rows) in enumerate(data_sections.items()):
-            worksheet = workbook.add_worksheet(key)
+            # worksheet max length name is 31 so need to truncate
+            worksheet_name = worksheet_names[index]
```

Comment:
Are dictionaries ordered? I am under assumption that they are not by default, but such ordereddict maybe exists.
Just wondering if this approach could technically produce incorrect names?

Reply:
Good question. If items() and keys() are called when no changes have been made to the dictionary then they should return same order. However, this is maybe not nicest way and there may be errors if someone does later decide to do changes on the dictionary between the calls. I can try to come up with a nicer solution on Friday if this can wait until then.

Reply:
New solution seems nicer, the original worksheetname is there, and the replacement, so there won't be disparencies in the comparison :+1: